idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,1160 @@
+                for idx3 in self.bidding_set[2]:
+
+                    aux_dict[model_sets[self.bidding_set[0]]] = idx1
+                    aux_dict[model_sets[self.bidding_set[1]]] = idx2
+                    aux_dict[model_sets[self.bidding_set[2]]] = idx3
+
+                    energy_price[idx1,idx2,idx3] = price_forecasts[aux_dict['Generators']][aux_dict['LMP Scenarios']][aux_dict['Time']]
+
+        return
+
+    def _assemble_bids(self):
+
+        '''
+        This methods extract the bids out of the stochastic programming model and
+        organize them.
+
+        Arguments:
+
+        Returns:
+            bids: the bid we computed. It is a dictionary that has this structure
+            {t: {gen:{power: cost}}}.
+        '''
+
+        model_sets = self.bidding_model_object.indices
+        model_sets_inverse = {}
+        for s, n in model_sets.items():
+            model_sets_inverse[n] = s
+
+        # unpack things
+        model_sets = self.bidding_model_object.indices
+        set_ls = list(model_sets.keys())
+
+        power_output = self.bidding_model_object.power_output
+        energy_price = self.model.energy_price
+
+        bids = {}
+        def dfs_extract_bids(set_idx, indices, bidding_indices, t, gen):
+
+            # traveled all the sets, and now we can add the objective
+            if set_idx == len(set_ls):
+                power = round(pyo.value(power_output[tuple(indices)]),2)
+                marginal_cost = round(pyo.value(energy_price[tuple(bidding_indices)]),2)
+
+                if t not in bids:
+                    bids[t] = {}
+
+                if gen not in bids[t]:
+                    bids[t][gen] = {}
+
+                # add without duplicate
+                if power in bids[t][gen]:
+                    bids[t][gen][power] = min(bids[t][gen][power], marginal_cost)
+                else:
+                    bids[t][gen][power] = marginal_cost
+
+                return
+
+            for idx in set_ls[set_idx]:
+
+                if model_sets[set_ls[set_idx]] == 'Time':
+                    t = idx
+                if model_sets[set_ls[set_idx]] == 'Generators':
+                    gen = idx
+
+                indices.append(idx)
+                if model_sets[set_ls[set_idx]] in self.bidding_set_names:
+                    bidding_indices.append(idx)
+
+                # recursion
+                dfs_extract_bids(set_idx + 1, indices, bidding_indices, t, gen)
+
+                # backtrack
+                indices.pop()
+                if model_sets[set_ls[set_idx]] in self.bidding_set_names:
+                    bidding_indices.pop()
+
+        dfs_extract_bids(set_idx = 0, \
+                         indices = [], \
+                         bidding_indices = [], \
+                         t = None,\
+                         gen = None)
+
+        for t in model_sets_inverse['Time']:
+            for gen in model_sets_inverse['Generators']:
+
```

Comment:
The comments in this section of nested for loops are really helpful. My suggestion is to comment the rest of the class with similar detail.